webapi/src/main/scala/org/knora/webapi/responders/v2/SearchResponderV2.scala

```diff
@@ -936,139 +1054,412 @@ class SearchResponderV2 extends ResponderWithStandoffV2 {
-                        Seq(
-                            StatementPattern.makeExplicit(subj = stringVar, pred = IriRef(OntologyConstants.KnoraBase.MatchesTextIndex.toSmartIri), XsdLiteral(searchTerms.combineSearchTermsWithLogicalAnd, OntologyConstants.Xsd.String.toSmartIri))
-                        )
+            // A variable representing the standoff link tag.
+            val standoffTagVar = functionCallExpression.getArgAsQueryVar(pos = 1)
+
+            typeInspectionResult.getTypeOfEntity(standoffTagVar) match {
+                case Some(NonPropertyTypeInfo(typeIri)) if typeIri.toString == OntologyConstants.KnoraApiV2WithValueObjects.StandoffTag => ()
+                case _ => throw GravsearchException(s"The second argument of ${functionIri.toSparql} must represent a knora-api:StandoffTag")
+            }
+
+            val linkTargetEntity = functionCallExpression.args(2) match {
+                case queryVar: QueryVariable => queryVar
+                case iriRef: IriRef => iriRef
+                case _ => throw GravsearchException(s"The third argument of ${functionIri.toSparql} must be a variable or IRI")
+            }
+
+            val statementsForTargetResource: Seq[QueryPattern] = typeInspectionResult.getTypeOfEntity(linkTargetEntity) match {
+                case Some(nonPropertyTpeInfo: NonPropertyTypeInfo) if OntologyConstants.KnoraApi.isKnoraApiV2Resource(nonPropertyTpeInfo.typeIri) =>
+
+                    // process the entity representing the target of the link
+                    createAdditionalStatementsForNonPropertyType(nonPropertyTpeInfo, linkTargetEntity)
+
+                case _ => throw GravsearchException(s"The third argument of ${functionIri.toSparql} must represent a knora-api:Resource")
+            }
+
+            val hasStandoffLinkToIriRef = IriRef(OntologyConstants.KnoraBase.HasStandoffLinkTo.toSmartIri)
+
+            // Generate statements linking the source resource and the standoff tag to the target resource.
+            val linkStatements = Seq(
+                StatementPattern.makeExplicit(subj = linkSourceEntity, pred = hasStandoffLinkToIriRef, obj = linkTargetEntity),
+                StatementPattern.makeInferred(subj = standoffTagVar, pred = IriRef(OntologyConstants.KnoraBase.StandoffTagHasLink.toSmartIri), obj = linkTargetEntity)
+            )
+
+            // Generate statements matching the link value that describes the standoff link between the source and target resources.
+            val statementsForLinkValue: Seq[StatementPattern] = generateStatementsForLinkValue(
+                linkSource = linkSourceEntity,
+                linkPred = hasStandoffLinkToIriRef,
+                linkTarget = linkTargetEntity
+            )
+
+            TransformedFilterPattern(
+                None, // FILTER has been replaced with statements
+                linkStatements ++ statementsForLinkValue ++ statementsForTargetResource
```

Comment:
Do you think we have to care about redundant triples?

```sparql
PREFIX knora-api: <http://api.knora.org/ontology/knora-api/v2#>
PREFIX standoff: <http://api.knora.org/ontology/standoff/v2#>
PREFIX anything: <http://0.0.0.0:3333/ontology/0001/anything/v2#>

CONSTRUCT {
    ?thing knora-api:isMainResource true .
    ?thing anything:hasText ?text .
    
} WHERE {
    ?thing a anything:Thing .
    ?thing anything:hasText ?text .
    ?text knora-api:textValueHasStandoff ?standoffTag .
    ?standoffTag a knora-api:StandoffLinkTag .
    FILTER knora-api:standoffLink(?thing, ?standoffTag, ?otherThing)
    ?otherThing a anything:Thing .
    
}
```

prequery:

```sparql
 SELECT DISTINCT ?thing (GROUP_CONCAT(DISTINCT(?otherThing); SEPARATOR='') AS ?otherThing__Concat) (GROUP_CONCAT(DISTINCT(?text); SEPARATOR='') AS ?text__Concat) (GROUP_CONCAT(DISTINCT(?otherThing__LinkValue); SEPARATOR='') AS ?otherThing__LinkValue__Concat)
 WHERE {
 ?thing <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/knora-base#Resource> .
 GRAPH <http://www.ontotext.com/explicit> {
     ?thing <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
 }
 ?thing <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/0001/anything#Thing> .
 ?thing <http://www.knora.org/ontology/0001/anything#hasText> ?text .
 GRAPH <http://www.ontotext.com/explicit> {
     ?text <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
 }
 ?text <http://www.knora.org/ontology/knora-base#valueHasStandoff> ?standoffTag .
 ?standoffTag <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/knora-base#StandoffLinkTag> .
 ?otherThing <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/knora-base#Resource> .
 GRAPH <http://www.ontotext.com/explicit> {
     ?otherThing <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
 }
 ?otherThing <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/0001/anything#Thing> .
 GRAPH <http://www.ontotext.com/explicit> {
     ?thing <http://www.knora.org/ontology/knora-base#hasStandoffLinkTo> ?otherThing .
 }
 ?standoffTag <http://www.knora.org/ontology/knora-base#standoffTagHasLink> ?otherThing .
 ?thing <http://www.knora.org/ontology/knora-base#hasStandoffLinkToValue> ?otherThing__LinkValue .
 GRAPH <http://www.ontotext.com/explicit> {
     ?otherThing__LinkValue <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/knora-base#LinkValue> .
 }
 GRAPH <http://www.ontotext.com/explicit> {
     ?otherThing__LinkValue <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
 }
 GRAPH <http://www.ontotext.com/explicit> {
     ?otherThing__LinkValue <http://www.w3.org/1999/02/22-rdf-syntax-ns#subject> ?thing .
 }
 GRAPH <http://www.ontotext.com/explicit> {
     ?otherThing__LinkValue <http://www.w3.org/1999/02/22-rdf-syntax-ns#object> ?otherThing .
 }
 ?otherThing <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/knora-base#Resource> .
 GRAPH <http://www.ontotext.com/explicit> {
     ?otherThing <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
 }
 }
 GROUP BY ?thing
 ORDER BY ASC(?thing)
 LIMIT 25
```

Note that 

```
 ?otherThing <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/knora-base#Resource> .
 GRAPH <http://www.ontotext.com/explicit> {
     ?otherThing <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
 }
``` is redundant. This is because there are two statements involving `?otherThing`.

Could we simply remove duplicates from the sequence of statements before creating the WHERE clause for the prequery?

Reply:
`createAdditionalStatementsForNonPropertyType ` is called twice: once from within `handleStandoffLinkFunction` for `?otherThing` and once because of ` ?otherThing a anything:Thing .`. We call it from within `handleStandoffLinkFunction` because there could be no statement ` ?otherThing a anything:Thing .` and then the logic would not work correctly.

Reply:
I guess this is because we look at statements one by one.

Reply:
How about if `createAdditionalStatementsForNonPropertyType` maintains a mutable set of statements that has already generated, so it can avoid generating them redundantly?

Reply:
Could we do that as an optimization in the refactoring we have in mind anyway?

Then we could merge this PR, so it does not block you from what you have to do.

Reply:
OK!